### PR TITLE
add hyperparameter optimalization tutorial

### PR DESCRIPTION
**Pull request by ErlendHov82**

### CODE REVIEW

File: hyper-para-opt/content.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+MFlux makes keeping track of and comparing such test runs easy.
+
+_Running the experiment in this tutorial will be quick and easy for the human, but the machine will have to work a long time, 
+likely several hours._
+
+## Prerequisites
+Create a new environment and run the following installations
+`conda install python=3.7`
+`pip install numpy==1.17 tensorflow==2 mlflow==1.3 boto3 keras==2.3 mflux_ai==0.5.1 scikit-learn==0.21`
```
Comment (iver56) [inline hyper-para-opt/content.md]: ~mflux_ai==0.5.1~ `mflux-ai>=0.5.3`

File: hyper-para-opt/content.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+target array from the form `[7]` to the form `[0, 0, 0, 0, 0, 0, 0, 1, 0, 0]`.
+
+
+## The Network
+We define a function that will build the network, making it easy to tweak certain hyperparameters. We use two layers of 
+convolutions (knowing in advance that this is a reasonable choice for the task) with pooling after each and one dense layer 
+at the end. We add dropout after each of these and flatten the output from the convolutions beforehand.
+
+The model needs to be logged to MFlux of course, but we also log a set of tags. These are the hyperparameters in question, 
+so we can easily see in the MFlux tracking upi what structure a particularly successful network has. Logging the history
```
Comment (iver56) [inline hyper-para-opt/content.md]: ~upi~ _UI_

File: hyper-para-opt/content.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+```
+Notice: We are choosing a random subset of 6000 validation images for every network we train. This may sound like a lot, but 
+most networks will score somewhere above 99% accuracy. This means we have less than 60 "tricky" images per validation set. 
+At this size, randomness can make itself know. If a network scores better on the validation set than the training set, it may 
+be that the validation set simply contains few hard cases.
+
+
+## Optimizing
+Now we can call this function with various inputs to look for the best setup. Training one model for 30 epochs takes around 
+10 minutes, depending on the computer it is run on. If we have 5 hyperparameters we are intrested in
```
Comment (iver56) [inline hyper-para-opt/content.md]: _interested_

**Review by iver56 — COMMENTED** [review]
What I really want to see in this tutorial is how to use tools like hyperopt/keras-tuner/hyperas/nevergrad (choose one) to tune hyperparameters _automatically_

File: hyper-para-opt/content.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+## Going further
+The best honest entries in the Kaggle comptition currently score around 99.7 % (on the unseen part of the data, the way it is split
+on Kaggle). Here are some suggestions for how to get closer to this score:
+* Obtain more powerful hardware and use GPU computing (for instance with CUDA). Speeding up training will let us do more 
+involved searches.
+* Use a smarter approach for the search +++
+* Look into more advanced network techniques like batch normalization, a deeper network or swapping the pooling layers 
+for convolutional layers.
+* Use a callback during training to keep the best model.
+* Apply tweaks like strech and rotation to the data to artificially increase the dataset in size.
```
Comment (iver56) [inline hyper-para-opt/content.md]: stretch

File: hyper-para-opt/content.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+* Use an average of many good networks.
+* Investigate the mislabled output images for ideas on what the network is struggling with.
```
Comment (iver56) [inline hyper-para-opt/content.md]: This is out of scope for this tutorial

File: hyper-para-opt/content.md [hunk 6]
```diff
@@ -1,150 +1,140 @@
...
-    return oh_target
+import mlflow.keras
+from hyperas import optim
+from hyperas.distributions import uniform, quniform
+from hyperopt import Trials, STATUS_OK, tpe
+```
 
+Initialize MFlux and set an experiment to store the models we will train in:
+```python
+mflux_ai.init("UC9VLr8l5lp6mgyt8aJ1MA")
```
Comment (iver56) [inline hyper-para-opt/content.md]: Replace this key, please

File: hyper-para-opt/content.md [hunk 7]
```diff
@@ -1,150 +1,140 @@
...
+                'batch_size': bs}
         mlflow.set_tags(tags)
-        for i in range(0, epochs, 1):
+        for i in range(epochs):
             mlflow.log_metric(key="val_acc", value=history.history["val_accuracy"][i], step=i)
             mlflow.log_metric(key="acc", value=history.history["accuracy"][i], step=i)
-    error = model.evaluate(val, val_target, verbose=0)
-    return model, error
+    score, acc = model.evaluate(X_test, Y_test, verbose=0)
+    return {'loss': -acc, 'status': STATUS_OK, 'model': model}
```
Comment (iver56) [inline hyper-para-opt/content.md]: Remove `, 'model': model` here, as that is a known source of memory leakage

File: hyper-para-opt/content.md [hunk 8]
```diff
@@ -1,150 +1,140 @@
...
+X_train, Y_train, X_test, Y_test = get_data()
+print("Evalutation of best performing model:")
+print(best_model.evaluate(X_test, Y_test, verbose=0))
 ```
 
-## Choosing a model
+
+## Inspecting the Models
+All the trained models have now been logged to MFlux, with the tags telling us what the hyperparameters are set to and a
+history of accuray and valuation accuracy per epoch.
```
Comment (iver56) [inline hyper-para-opt/content.md]: _accuracy_
_validation_

File: hyper-para-opt/content.md [hunk 9]
```diff
@@ -1,150 +1,140 @@
...
+
 The main page for this experiment in MFlux will look something like this
 ![Experiment main page](images/experiment_main.png)
 
 We can sort valuation accuracy by descending to see the models that perform best on images not seen during training
 ![Experiment sort](images/experiment_sort.png)
 
 By clicking on a run (under "Date") we can see more details about it. This is the graph for valuation accuracy over time for
```
Comment (iver56) [inline hyper-para-opt/content.md]: _validation_

File: hyper-para-opt/content.md [hunk 10]
```diff
@@ -1,150 +1,140 @@
...
-## Choosing a model
+
+## Inspecting the Models
+All the trained models have now been logged to MFlux, with the tags telling us what the hyperparameters are set to and a
+history of accuray and valuation accuracy per epoch.
+
 The main page for this experiment in MFlux will look something like this
 ![Experiment main page](images/experiment_main.png)
 
 We can sort valuation accuracy by descending to see the models that perform best on images not seen during training
```
Comment (iver56) [inline hyper-para-opt/content.md]: _validation_

**Review by iver56 — CHANGES_REQUESTED** [review]
Please replace instances of "MFlux" with "MFlux.ai"

**Review by iver56 — APPROVED** [review]
